service/storage_service.cc

```diff
@@ -940,39 +940,61 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
+    // and eventually, if any candidate for removal end up owning no tokens, it is removed
+    // from token_metadata.
+    std::unordered_map<token, inet_address> token_to_endpoint_map = get_token_metadata().get_token_to_endpoint();
+    std::unordered_set<inet_address> candidates_for_removal;
+
     for (auto t : tokens) {
         // we don't want to update if this node is responsible for the token and it has a later startup time than endpoint.
-        auto current_owner = tmptr->get_endpoint(t);
-        if (!current_owner) {
+        auto current = token_to_endpoint_map.find(t);
+        if (current == token_to_endpoint_map.end()) {
             slogger.debug("handle_state_normal: New node {} at token {}", endpoint, t);
             owned_tokens.insert(t);
-        } else if (endpoint == *current_owner) {
-            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, *current_owner, t);
+            continue;
+        }
+        auto current_owner = current->second;
+        if (endpoint == current_owner) {
+            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, current_owner, t);
             // set state back to normal, since the node may have tried to leave, but failed and is now back up
             owned_tokens.insert(t);
-        } else if (_gossiper.compare_endpoint_startup(endpoint, *current_owner) > 0) {
-            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, *current_owner, t);
+            token_to_endpoint_map.erase(current);
+        } else if (_gossiper.compare_endpoint_startup(endpoint, current_owner) > 0) {
+            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, current_owner, t);
             owned_tokens.insert(t);
+            slogger.info("handle_state_normal: remove endpoint={} token={}", current_owner, t);
             // currentOwner is no longer current, endpoint is.  Keep track of these moves, because when
             // a host no longer has any tokens, we'll want to remove it.
-            std::multimap<inet_address, token> ep_to_token_copy = get_token_metadata().get_endpoint_to_token_map_for_reading();
-            auto rg = ep_to_token_copy.equal_range(*current_owner);
-            for (auto it = rg.first; it != rg.second; it++) {
-                if (it->second == t) {
-                    slogger.info("handle_state_normal: remove endpoint={} token={}", *current_owner, t);
-                    ep_to_token_copy.erase(it);
+            token_to_endpoint_map.erase(current);
+            candidates_for_removal.insert(current_owner);
+            slogger.info("handle_state_normal: Nodes {} and {} have the same token {}. {} is the new owner", endpoint, current_owner, t, endpoint);
+        } else {
+            slogger.info("handle_state_normal: Nodes {} and {} have the same token {}. Ignoring {}", endpoint, current_owner, t, endpoint);
```

Comment:
Hmm, unfortunately I see we need to remove endpoint here too if it has no tokens.
See https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/105/artifact/logs-heavy.debug.002/1670131224377_replace_address_test.py%3A%3ATestReplaceAddress%3A%3Atest_replace_node_diff_ip_take_write%5Brbo_disabled%5D/node6.log
```
...
INFO  2022-12-04 05:20:09,267 [shard 0] storage_service - handle_state_normal: Nodes 127.0.18.5 and 127.0.18.6 have the same token -4248524798093784041. Ignoring 127.0.18.5
ERROR 2022-12-04 05:20:09,267 [shard 0] storage_service - endpoint=127.0.18.5 is not marked for removal but owns no tokens, at: 0xd752170 0x13cdab05 0x13cda85d 0x13cdcc73 0x13cdd6ff 0x12d4b6b2 0x12d4be18 0x10ab2667 0xd9c8d76 0x12eadad9 0x12eb50ce 0x12eb90bf 0x12eb6d3c 0x12bd258e 0x12bcfd86 0xd7e6f6c 0xd87ba45 0xd7e5168 /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/b551cd254c6daf07014267950fcdd17aaf19acf0/libreloc/libc.so.6+0x2750f /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/b551cd254c6daf07014267950fcdd17aaf19acf0/libreloc/libc.so.6+0x275c8 0xd6e8964
```

Decoded:
```
void seastar::backtrace<seastar::current_backtrace_tasklocal()::$_3>(seastar::current_backtrace_tasklocal()::$_3&&) at ./build/debug/seastar/./seastar/include/seastar/util/backtrace.hh:59
seastar::current_backtrace_tasklocal() at ./build/debug/seastar/./seastar/src/util/backtrace.cc:86
seastar::current_tasktrace() at ./build/debug/seastar/./seastar/src/util/backtrace.cc:137
seastar::current_backtrace() at ./build/debug/seastar/./seastar/src/util/backtrace.cc:170
log_error_and_backtrace(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at ./build/debug/seastar/./seastar/src/core/on_internal_error.cc:37
seastar::on_internal_error_noexcept(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at ./build/debug/seastar/./seastar/src/core/on_internal_error.cc:60
service::storage_service::handle_state_normal(gms::inet_address) at ./service/storage_service.cc:985
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/coroutine:244
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:120
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/debug/seastar/./seastar/src/core/reactor.cc:2486
seastar::reactor::run_some_tasks() at ./build/debug/seastar/./seastar/src/core/reactor.cc:2923
seastar::reactor::do_run() at ./build/debug/seastar/./seastar/src/core/reactor.cc:3092
seastar::reactor::run() at ./build/debug/seastar/./seastar/src/core/reactor.cc:2975
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ./build/debug/seastar/./seastar/src/core/app-template.cc:265
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at ./build/debug/seastar/./seastar/src/core/app-template.cc:156
scylla_main(int, char**) at ./main.cc:526
```